core/slice.go

```diff
@@ -0,0 +1,484 @@
+	"github.com/spruce-solutions/go-quai/consensus"
+	"github.com/spruce-solutions/go-quai/core/rawdb"
+	"github.com/spruce-solutions/go-quai/core/types"
+	"github.com/spruce-solutions/go-quai/core/vm"
+	"github.com/spruce-solutions/go-quai/ethclient/quaiclient"
+	"github.com/spruce-solutions/go-quai/ethdb"
+	"github.com/spruce-solutions/go-quai/log"
+	"github.com/spruce-solutions/go-quai/params"
+)
+
+type Slice struct {
+	hc *HeaderChain
+
+	config *params.ChainConfig
+	engine consensus.Engine
+
+	quit chan struct{} // slice quit channel
+
+	domClient    *quaiclient.Client   // domClient is used to check if a given dominant block in the chain is canonical in dominant chain.
+	subClients   []*quaiclient.Client // subClinets is used to check is a coincident block is valid in the subordinate context
+	futureBlocks *lru.Cache
+
+	wg sync.WaitGroup // slice processing wait group for shutting down
+}
+
+func NewSlice(db ethdb.Database, chainConfig *params.ChainConfig, domClientUrl string, subClientUrls []string, engine consensus.Engine, vmConfig vm.Config) (*Slice, error) {
+	sl := &Slice{
+		config: chainConfig,
+		engine: engine,
+	}
+
+	futureBlocks, _ := lru.New(maxFutureBlocks)
+	sl.futureBlocks = futureBlocks
+
+	var err error
+	sl.hc, err = NewHeaderChain(db, engine, chainConfig, vmConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// only set the domClient if the chain is not prime
+	if types.QuaiNetworkContext != params.PRIME {
+		sl.domClient = MakeDomClient(domClientUrl)
+	}
+
+	sl.subClients = make([]*quaiclient.Client, 3)
+	// only set the subClients if the chain is not region
+	if types.QuaiNetworkContext != params.ZONE {
+		go func() {
+			sl.subClients = MakeSubClients(subClientUrls)
+		}()
+	}
+
+	return sl, nil
+}
+
+// HeaderChain retrieves the headerchain.
+func (sl *Slice) HeaderChain() *HeaderChain {
+	return sl.hc
+}
+
+// MakeDomClient creates the quaiclient for the given domurl
+func MakeDomClient(domurl string) *quaiclient.Client {
+	if domurl == "" {
+		log.Crit("dom client url is empty")
+	}
+	domClient, err := quaiclient.Dial(domurl)
+	if err != nil {
+		log.Crit("Error connecting to the dominant go-quai client", "err", err)
+	}
+	return domClient
+}
+
+// MakeSubClients creates the quaiclient for the given suburls
+func MakeSubClients(suburls []string) []*quaiclient.Client {
+	subClients := make([]*quaiclient.Client, 3)
+	for i, suburl := range suburls {
+		if suburl == "" {
+			log.Warn("sub client url is empty")
+		}
+		subClient, err := quaiclient.Dial(suburl)
+		if err != nil {
+			log.Crit("Error connecting to the subordinate go-quai client for index", "index", i, " err ", err)
+		}
+		subClients[i] = subClient
+	}
+	return subClients
+}
+
+// Config retrieves the slice's chain configuration.
+func (sl *Slice) Config() *params.ChainConfig { return sl.config }
+
+// Engine retrieves the header chain's consensus engine.
+func (sl *Slice) Engine() consensus.Engine { return sl.engine }
+
+// Append
```

Comment:
Expounding on the documentation would be nice :)

Not a blocker, but if you have the time.